src/rigid.rs

```diff
@@ -0,0 +1,265 @@
+
+impl<T: Float + ApproxEq<T>, Src, Dst> TypedRigidTransform3D<T, Src, Dst> {
+    /// Construct a new rigid transformation, where the `rotation` applies first
+    #[inline]
+    pub fn new(rotation: TypedRotation3D<T, Src, Dst>, translation: TypedVector3D<T, Dst>) -> Self {
+        Self {
+            rotation,
+            translation,
+        }
+    }
+
+    /// Construct an identity transform
+    #[inline]
+    pub fn identity() -> Self {
+        Self {
+            rotation: TypedRotation3D::identity(),
+            translation: TypedVector3D::zero(),
+        }
+    }
+
+    /// Construct a new rigid transformation, where the `translation` applies first
+    #[inline]
+    pub fn new_from_reversed(
+        translation: TypedVector3D<T, Src>,
+        rotation: TypedRotation3D<T, Src, Dst>,
+    ) -> Self {
+        // T * R
+        //   = (R * R^-1) * T * R
+        //   = R * (R^-1 * T * R)
+        //   = R * T'
+        //
+        // T' = (R^-1 * T * R) is also a translation matrix
+        // It is equivalent to the translation matrix obtained by rotating the
+        // translation by R
+
+        let translation = rotation.rotate_vector3d(&translation);
+        Self {
+            rotation,
+            translation,
+        }
+    }
+
+    #[inline]
+    pub fn from_rotation(rotation: TypedRotation3D<T, Src, Dst>) -> Self {
+        Self {
+            rotation,
+            translation: TypedVector3D::zero(),
+        }
+    }
+
+    #[inline]
+    pub fn from_translation(translation: TypedVector3D<T, Dst>) -> Self {
+        Self {
+            translation,
+            rotation: TypedRotation3D::identity(),
+        }
+    }
+
+    /// Decompose this into a position and an orientation to be applied in the opposite order
+    #[inline]
+    pub fn decompose_reversed(&self) -> (TypedVector3D<T, Src>, TypedRotation3D<T, Src, Dst>) {
+        // self = R * T
+        //      = R * T * (R^-1 * R)
+        //      = (R * T * R^-1) * R)
+        //      = T' * R
+        //
+        // T' = (R^ * T * R^-1) is T rotated by R^-1
+
+        let translation = self.rotation.inverse().rotate_vector3d(&self.translation);
+        (translation, self.rotation)
+    }
+
+    /// Returns the multiplication of the two transforms such that
+    /// other's transformation applies after self's transformation.
+    ///
+    /// i.e., this produces `self * other` in row-vector notation
+    #[inline]
+    pub fn post_mul<Dst2>(
+        &self,
+        other: &TypedRigidTransform3D<T, Dst, Dst2>,
+    ) -> TypedRigidTransform3D<T, Src, Dst2> {
+        // self = R1 * T1
+        // other = R2 * T2
+        // result = R1 * T1 * R2 * T2
+        //        = R1 * T1 * (R1^-1 * R1) * R2 * T2
+        //        = (R1 * T1 * R1^-1) * R1 * R2 * T2
+        //        = T' * R' * T2
+        //        = (R' * R'^-1) * T' * R' * T2
+        //        = R' * (R'^-1 * T' * R') * T2
+        //        = R' * T'' * T2
+        //        = R * T''
```

Comment:
How do you get this last step from the above one?

Reply:
er, typo, it should be R'